### ISSUE TITLE
The 128-bit of interface dispatch cache entry and indirection cell's stub/cache mismatch the PalInterlockedCompareExchange128's atomically requirements

### ISSUE THREAD

**Issue by shushanhf** [dotnet/runtime#109276]
### Description

https://github.com/dotnet/runtime/pull/108364#issuecomment-2437185184

I have some question about the CAS-128bit when updating and loading the 128-bit data.
https://github.com/dotnet/runtime/blob/5d09a8f94c72ca4ef0a9c79eb9c58d06198e3ba9/src/coreclr/nativeaot/Runtime/amd64/StubDispatch.S#L30-L32

Are these codes related with the CAS-128-bit data?
If yes, is there some problem if the 128-bit data are loaded by two instructions rather than a 128-bit load instruction? When finished the first 64-bit pointer loading, the CAS-128-bit update the whole 128-bit date, then the second loading will get a new value but now the old value.

### Reproduction Steps

When finished the first 64-bit pointer loading, the CAS-128-bit update the whole 128-bit date, then the second loading will get a new value but now the old value.

### Expected behavior

should load the 128-bit date by one instruction atomically.

### Actual behavior

But now these code will hit the wrong value at the very low possibility.

**Comment by am11** [comment]
Is this question is about how it's currently handled on x64 (without cmpxchg16b or atomic guarantees etc.), not loongarch itself, right?

**Comment by shushanhf** [comment]
> Is this question about how it's currently handled on x64 (without cmpxchg16b or atomic guarantees etc.), 

No, for ARM64 and AMD64, there is potential getting a wrong value when loading the 128-bit of interface dispatch cache entry and indirection cell's stub/cache.

> not loongarch itself, right?

yes, for all the targets.

**Comment by MichalStrehovsky** [comment]
I think the only possible transition here is from null to a non-null value (we'll not update a non-null value to anything else):

https://github.com/dotnet/runtime/blob/cca022b6212f33adc982630ab91469882250256c/src/coreclr/nativeaot/Runtime/CachedInterfaceDispatch.cpp#L69-L81

Now, I'm not an expert on memory models, but if the first 64-bit read saw a non-null value from the CAS-128, the second 64-bit read should also see the new value (and not the null).

**Comment by shushanhf** [comment]
```
|                 Thread-1: reading the 128-bit data.                         |    Thread-2: updating 128-bit data by CAS-128.
| clock N:     get the first 64-bit date.                                         |    After Thead-1 getted the first 64-bit data,
| clock N+1:                                                                             |    the Thread-2  updated the whole 128-bit data by CAS-128. 
| clock N+2:  when the Thread-1 reading the second 64-bit
|                   it will get the new value.
|                   The Thread-1 will hit a wrong value that not expected value.
```

**Comment by MichalStrehovsky** [comment]
Correct me if I'm wrong, but since the CAS-128 is only used to flip both values from zero to non-zero, Thread-1 will either read 0 in the first 64-bit read (in which case it will not do the second read) or read the value that CAS-128 placed there and the second read will read a value placed there by the same CAS-128

We should never see torn state because CAS-128 only flips from zero to non-zero per the above comment ("`cache entry updates  (fFailOnNonNull == true) require that the existing values in the location are both NULL`"). It never flips non-zero value to a different non-zero value.

**Comment by shushanhf** [comment]
> Correct me if I'm wrong, but since the CAS-128 is only used to flip both values from zero to non-zero, Thread-1 will either read 0 in the first 64-bit read (in which case it will not do the second read) or read the value that CAS-128 placed there and the second read will read a value placed there by the same CAS-128
> 
> We should never see torn state because CAS-128 only flips from zero to non-zero per the above comment ("`cache entry updates (fFailOnNonNull == true) require that the existing values in the location are both NULL`"). It never flips non-zero value to a different non-zero value.

I just focus on the instructions by the ISA's point and I'm not familiar with the detail usage of the `UpdatePointerPairAtomically`.
What you means that the `UpdatePointerPairAtomically` will be called by each interface dispatch cache entry only one time, is it right ?

**Comment by MichalStrehovsky** [comment]
> > Correct me if I'm wrong, but since the CAS-128 is only used to flip both values from zero to non-zero, Thread-1 will either read 0 in the first 64-bit read (in which case it will not do the second read) or read the value that CAS-128 placed there and the second read will read a value placed there by the same CAS-128
> > We should never see torn state because CAS-128 only flips from zero to non-zero per the above comment ("`cache entry updates (fFailOnNonNull == true) require that the existing values in the location are both NULL`"). It never flips non-zero value to a different non-zero value.
> 
> I just focus on the instructions by the ISA's point and I'm not familiar with the detail usage of the `UpdatePointerPairAtomically`. What you means that the `UpdatePointerPairAtomically` will be called by each interface dispatch cache entry only one time, is it right ?

Yes, I believe the assembly code you have in top-post only consumes data that was updated with `UpdatePointerPairAtomically` where fFailOnNonNull == true. This use of `UpdatePointerPairAtomically` could likely be replaced by two 64-bit writes (first write the second value, then the first one). It probably doesn't _need_ to be a 128-bit CAS.

However the other use of `UpdatePointerPairAtomically` with `fFailOnNonNull == false` looks like it really needs a 128-bit CAS since that one reads both values unconditionally and we do allow arbitary overwrites. The output of that is not read by the above assembly code. It's part of cached interface dispatch logic.

**Comment by heiher** [comment]
> Correct me if I'm wrong, but since the CAS-128 is only used to flip both values from zero to non-zero, Thread-1 will either read 0 in the first 64-bit read (in which case it will not do the second read) or read the value that CAS-128 placed there and the second read will read a value placed there by the same CAS-128
> 
> We should never see torn state because CAS-128 only flips from zero to non-zero per the above comment ("`cache entry updates (fFailOnNonNull == true) require that the existing values in the location are both NULL`"). It never flips non-zero value to a different non-zero value.

The loads can be reordered after loads on most [weak memory model architectures](https://en.wikipedia.org/wiki/Memory_ordering#In_symmetric_multiprocessing_(SMP)_microprocessor_systems). It is possible for the first 64 bits to be loaded with a non-zero value while the second 64 bits are loaded with a zero value, unless there is an acquire semantic memory barrier between the two loads.

**Comment by shushanhf** [comment]
> > > Correct me if I'm wrong, but since the CAS-128 is only used to flip both values from zero to non-zero, Thread-1 will either read 0 in the first 64-bit read (in which case it will not do the second read) or read the value that CAS-128 placed there and the second read will read a value placed there by the same CAS-128
> > > We should never see torn state because CAS-128 only flips from zero to non-zero per the above comment ("`cache entry updates (fFailOnNonNull == true) require that the existing values in the location are both NULL`"). It never flips non-zero value to a different non-zero value.
> > 
> > 
> > I just focus on the instructions by the ISA's point and I'm not familiar with the detail usage of the `UpdatePointerPairAtomically`. What you means that the `UpdatePointerPairAtomically` will be called by each interface dispatch cache entry only one time, is it right ?
> 
> Yes, I believe the assembly code you have in top-post only consumes data that was updated with `UpdatePointerPairAtomically` where fFailOnNonNull == true. This use of `UpdatePointerPairAtomically` could likely be replaced by two 64-bit writes (first write the second value, then the first one). It probably doesn't _need_ to be a 128-bit CAS.

Is it here that the `fFailOnNonNull` is passed by true,
https://github.com/dotnet/runtime/blob/cca022b6212f33adc982630ab91469882250256c/src/coreclr/nativeaot/Runtime/CachedInterfaceDispatch.cpp#L144

So the Line:92-Line96 below will not excuted. here it probably doesn't _need_ to be a 128-bit CAS, is it right ?
https://github.com/dotnet/runtime/blob/cca022b6212f33adc982630ab91469882250256c/src/coreclr/nativeaot/Runtime/CachedInterfaceDispatch.cpp#L91-L96

> However the other use of `UpdatePointerPairAtomically` with `fFailOnNonNull == false` looks like it really needs a 128-bit CAS since that one reads both values unconditionally and we do allow arbitary overwrites. The output of that is not read by the above assembly code. It's part of cached interface dispatch logic.

But here https://github.com/dotnet/runtime/blob/cca022b6212f33adc982630ab91469882250256c/src/coreclr/nativeaot/Runtime/CachedInterfaceDispatch.cpp#L159
the `fFailOnNonNull` is passed by false, this case will be different with the above case and this case might hit  getting two different version 64-bit data while the two 64-bit data are not paired, is is right ?

**Comment by heiher** [comment]
Here’s a test case reproduced on Apple M-series and Huawei Kunpeng920 CPUs. It verifies that on architectures with weak memory models allowing load and load reordering, atomic stores on one side to two parts of the same cache line cannot guarantee that non-atomic load reads from the same version of value.

```c
/*
 * memory-order-d64-same-cacheline.c
 * Dual 64-bit memory sequential test case in the same cache line.
 * gcc -O3 -o memory-order-d64-same-cacheline memory-order-d64-same-cacheline.c
 */

#include <stdio.h>
#include <stdint.h>
#include <stdlib.h>
#include <pthread.h>

static uint64_t d64[2] __attribute__((aligned(16)));

static void *
writer_handler (void *data)
{
    uint64_t i;
    __int128_t curr = *(__int128_t *) d64;

    for (i = 2; i < (UINT64_MAX - 10); i += 2) {
        __int128_t next = ((__int128_t) i + 1) << 64 | i;
        __sync_val_compare_and_swap ((volatile __int128_t *) d64, curr, next);
        curr = next;
    }

    printf ("SUCC\n");
    exit (0);

    return NULL;
}

int
main (int argc, char *argv[])
{
    pthread_t writer;
    int res = 0;

    *(volatile uint64_t *) &d64[1] = 1;
    *(volatile uint64_t *) &d64[0] = 0;

    res |= pthread_create (&writer, NULL, writer_handler, NULL);
    if (res < 0)
        return -1;

    for (;;) {
        uint64_t lo, hi;

        lo = *(volatile uint64_t *) &d64[0];
        hi = *(volatile uint64_t *) &d64[1];

        if (lo >= hi) {
            printf ("ASSERATION FAILED: %llu < %llu\n", lo, hi);
            return -2;
        }
    }

    return 0;
}
```

Specifically, in each iteration, the writer performs atomic stores of two 64-bit values into the same cache line, with each new version always greater than the previous one.

| Value version | High 64 bits | Low 64 bits |
| ------------------ | ---------------- | --------------- |
| 0                     | 1                  | 0                 |
| 1                     | 3                  | 2                 |
| 2                     | 5                  | 4                 |
| 3                     | 7                  | 6                 |
| 4                     | 9                  | 8                 |
| ...                    | ...                 | ...                |
| N                    | N * 2 + 1      | N * 2           |

The reader first loads the lower 64 bits and then the upper 64 bits. If the lower 64-bit value is greater than or equal to the upper 64-bit value, it indicates that reordering has occurred. The branch instruction lacks `acquire` semantics, so this is similarly applicable to the `(NULL, NULL)` -> `(!NULL, !NULL)` scenario in the current context.

```
Initial value: (NULL, NULL)

Writer                | Reader
(Non-NULL, Non-NULL)  |
                      | (Non-NULL) // Low 64 bits
                      | (NULL)     // High 64 bits
```

**Comment by MichalStrehovsky** [comment]
We no longer rely on 128-bit compare-exchange after #123252.

**Comment by am11** [comment]
These can be cleaned up it seems:

```sh
src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Unix.targets:241:      <!-- See the comment in PalInterlockedCompareExchange128 for details. -->

src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Unix.targets:242:      <NativeSystemLibrary Include="atomic" Condition="'$(_targetArchitecture)' == 'riscv64' or '$(_targetArchitecture)' == 'loongarch64'" />

src/coreclr/nativeaot/Runtime/unix/PalInline.h:98:FORCEINLINE uint8_t PalInterlockedCompareExchange128(_Inout_ int64_t volatile *pDst, int64_t iValueHigh, int64_t iValueLow, int64_t *pComparandAndResult)
```
